producer/writer/message_writer.go

```diff
@@ -0,0 +1,425 @@
+		msgID:             0,
+		queue:             list.New(),
+		consumerWriters:   make(map[string]consumerWriter),
+		acks:              newAckHelper(defaultAckMapSize),
+		cutOffNanos:       0,
+		cutOverNanos:      0,
+		isClosed:          false,
+		doneCh:            make(chan struct{}),
+		m:                 newMessageWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+		nowFn:             time.Now,
+	}
+}
+
+func (w *messageWriterImpl) Write(rd producer.RefCountedData) {
+	nowNanos := w.nowFn().UnixNano()
+	w.RLock()
+	isValid := w.isValidWriteWithLock(nowNanos)
+	w.RUnlock()
+	if !isValid {
+		return
+	}
+	rd.IncRef()
+	msg := w.mPool.Get()
+
+	w.Lock()
+	w.msgID++
+	meta := metadata{
+		shard: w.replicatedShardID,
+		id:    w.msgID,
+	}
+	msg.Reset(meta, rd)
+	w.acks.add(meta, msg)
+	w.writeWithLock(msg, nowNanos)
+	w.queue.PushBack(msg)
+	w.Unlock()
+
+	w.m.writeNew.Inc(1)
+}
+
+func (w *messageWriterImpl) isValidWriteWithLock(nowNanos int64) bool {
+	if w.cutOffNanos > 0 && nowNanos >= w.cutOffNanos {
+		w.m.writeAfterCutoff.Inc(1)
+		return false
+	}
+	if w.cutOverNanos > 0 && nowNanos < w.cutOverNanos {
+		w.m.writeBeforeCutover.Inc(1)
+		return false
+	}
+	return true
+}
+
+func (w *messageWriterImpl) writeWithLock(m *message, nowNanos int64) {
+	m.IncWriteTimesWithLock()
+	m.IncReads()
+	msg, isValid := m.Marshaler()
+	if !isValid {
+		m.DecReads()
+		return
+	}
+	written := false
+	for _, cw := range w.consumerWriters {
+		if err := cw.Write(msg); err != nil {
+			w.m.oneConsumerWriteError.Inc(1)
+			continue
+		}
+		written = true
+		w.m.writeSuccess.Inc(1)
+		break
+	}
+	m.DecReads()
+
+	if !written {
+		// Could not be written to any consumer, will retry later.
+		w.m.allConsumersWriteError.Inc(1)
+	}
+	m.SetRetryAtNanosWithLock(w.nextRetryNanos(m.WriteTimesWithLock(), nowNanos))
+}
+
+// TODO: make retry time strategy configurable.
+func (w *messageWriterImpl) nextRetryNanos(writeTimes int64, nowNanos int64) int64 {
+	half := w.backOffNanos / 2
+	backOff := half + rand.Int63n(half)
+	retryAtNanos := nowNanos + backOff
+	if writeTimes > 1 {
+		retryAtNanos += (writeTimes - 1) * w.backOffNanos
+	}
+	return retryAtNanos
+}
+
+func (w *messageWriterImpl) Ack(meta metadata) {
+	w.acks.ack(meta)
+}
+
+func (w *messageWriterImpl) Init() {
+	w.wg.Add(1)
+	go func() {
+		w.retryUnacknowledgedForever()
+		w.wg.Done()
+	}()
+}
+
+func (w *messageWriterImpl) retryUnacknowledgedForever() {
+	ticker := time.NewTicker(w.opts.MessageRetryBackoff())
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			w.retryUnacknowledged()
+		case <-w.doneCh:
+			return
+		}
+	}
+}
+
+func (w *messageWriterImpl) retryUnacknowledged() {
+	w.RLock()
+	e := w.queue.Front()
+	l := w.queue.Len()
+	w.RUnlock()
+	w.m.queueSize.Update(float64(l))
+	for e != nil {
+		now := w.nowFn()
+		nowNanos := now.UnixNano()
+		w.Lock()
+		e = w.retryBatchWithLock(e, nowNanos)
+		w.Unlock()
+		w.m.retryBatchLatency.Record(w.nowFn().Sub(now))
```

Comment:
Do you want to throttle here to give other threads a chance? similar logic in m3db and m3aggregator uses a const duration per element

Reply:
We could, not sure if super necessary though, my retry batch function currently breaks after encountering a message that needs to be written or passed ~10ms, so I feel that is already giving reasonable chances to other threads(new writes)